File: app/src/main/java/com/quran/labs/androidquran/QuranImportActivity.java
package com.quran.labs.androidquran;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import android.widget.Toast;

import com.quran.data.model.bookmark.BookmarkData;
import com.quran.labs.androidquran.presenter.QuranImportPresenter;
import com.quran.labs.androidquran.ui.util.ToastCompat;

import javax.inject.Inject;

public class QuranImportActivity extends AppCompatActivity implements
    ActivityCompat.OnRequestPermissionsResultCallback {
  private AlertDialog mDialog;
  @Inject QuranImportPresenter mPresenter;

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    QuranApplication quranApp = ((QuranApplication) getApplication());
    quranApp.refreshLocale(this, false);
    super.onCreate(savedInstanceState);
    quranApp.getApplicationComponent().inject(this);
  }

  @Override
  protected void onPause() {
    mPresenter.unbind(this);
    super.onPause();
  }

  @Override
  protected void onResume() {
    super.onResume();
    mPresenter.bind(this);
  }

  @Override
  protected void onDestroy() {
    mPresenter.unbind(this);
    if (mDialog != null) {
      mDialog.dismiss();
    }
    super.onDestroy();
  }

  public boolean isShowingDialog() {
    return mDialog != null;
  }

  public void showImportConfirmationDialog(final BookmarkData bookmarkData) {
    String message = getString(R.string.import_data_and_override,
        bookmarkData.getBookmarks().size(),
        bookmarkData.getTags().size());
    AlertDialog.Builder builder = new AlertDialog.Builder(this)
        .setMessage(message)
        .setPositiveButton(R.string.import_data,
            (dialog, which) -> mPresenter.importData(bookmarkData))
        .setNegativeButton(android.R.string.cancel, (dialog, which) -> finish())
        .setOnCancelListener(dialog -> finish());
    mDialog = builder.show();
  }

  public void showImportComplete() {
    ToastCompat.makeText(QuranImportActivity.this,
        R.string.import_successful, Toast.LENGTH_LONG).show();
    finish();
  }

  public void showError() {
    showErrorInternal(R.string.import_data_error);
  }

  public void showPermissionsError() {
    showErrorInternal(R.string.import_data_permissions_error);
  }

  private void showErrorInternal(int messageId) {
    AlertDialog.Builder builder = new AlertDialog.Builder(this)
        .setMessage(messageId)
        .setPositiveButton(android.R.string.ok, (dialog, which) -> finish());
    mDialog = builder.show();
  }

  @Override
  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
      @NonNull int[] grantResults) {
    mPresenter.onPermissionsResult(requestCode, grantResults);
  }
}

File: app/src/main/java/com/quran/labs/androidquran/QuranImportActivity.kt
package com.quran.labs.androidquran

import android.os.Bundle
import android.widget.Toast
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.quran.data.model.bookmark.BookmarkData
import com.quran.labs.androidquran.presenter.QuranImportPresenter
import com.quran.labs.androidquran.ui.util.ToastCompat
import javax.inject.Inject

class QuranImportActivity : AppCompatActivity(), ActivityCompat.OnRequestPermissionsResultCallback {

  private var alertDialog: AlertDialog? = null

  @Inject
  lateinit var presenter: QuranImportPresenter

  override fun onCreate(savedInstanceState: Bundle?) {
    val quranApp = application as QuranApplication
    quranApp.refreshLocale(this, false)
    super.onCreate(savedInstanceState)
    quranApp.applicationComponent.inject(this)
  }

  override fun onResume() {
    super.onResume()
    presenter.bind(this)
  }

  override fun onPause() {
    presenter.unbind(this)
    super.onPause()
  }

  override fun onDestroy() {
    presenter.unbind(this)
    alertDialog?.dismiss()
    super.onDestroy()
  }

  override fun onRequestPermissionsResult(
    requestCode: Int,
    permissions: Array<out String>,
    grantResults: IntArray
  ) {
    presenter.onPermissionsResult(requestCode, grantResults)
    super.onRequestPermissionsResult(requestCode, permissions, grantResults)
  }

  fun isShowingDialog(): Boolean = alertDialog != null

  fun showImportConfirmationDialog(bookmarkData: BookmarkData) {
    val dialogMessage = getString(
      R.string.import_data_and_override,
      bookmarkData.bookmarks.size,
      bookmarkData.tags.size
    )
    val builder = AlertDialog.Builder(this).apply {
      setMessage(dialogMessage)
      setPositiveButton(R.string.import_data) { _, _ ->
        presenter.importData(bookmarkData)
      }
      setNegativeButton(android.R.string.cancel) { _, _ -> finish() }
      setOnCancelListener { finish() }
    }
    alertDialog = builder.show()
  }

  fun showImportComplete() {
    ToastCompat.makeText(this, R.string.import_successful, Toast.LENGTH_LONG).show()
    finish()
  }

  fun showError() {
    showErrorInternal(R.string.import_data_error)
  }

  fun showPermissionsError() {
    showErrorInternal(R.string.import_data_permissions_error)
  }

  private fun showErrorInternal(@StringRes messageId: Int) {
    val builder = AlertDialog.Builder(this).apply {
      setMessage(messageId)
      setPositiveButton(android.R.string.ok) { _, _ -> finish() }
    }
    alertDialog = builder.show()
  }
}
